### PR TITLE
Shib details

### PR DESCRIPTION
**Pull request by Feverfew**
## What does this PR do?
This patch does the following:
1. Adds additional fields to the User model, which are present in the
Shibboleth callback.
2. Amends Shibboleth callback functions to add these fields to User
models created or updated in the callback.
3. `validate_shibboleth_callback()` function was created to keep
adherence to DRY, with corresponding tests.
4. Adds additional information in 400 responses when critical headers
are missing to aid debugging. See #3072 

## ✅ Pull Request checklist

- [x] Is this code complete?
- [x] Are tests done/modified?
- [x] Are docs written for this PR? (if applicable)

## 🚨 Is this a breaking change for API clients?
No
## :squirrel: Deploy notes
`python3 manage.py migrate dashboard`

### CODE REVIEW

File: backend/uclapi/dashboard/models.py [hunk 1]
```diff
@@ -20,10 +20,16 @@ class User(models.Model):
...
+    sn = models.CharField(max_length=100, default='')
     cn = models.CharField(max_length=100, unique=True)
     department = models.CharField(max_length=1000)
     employee_id = models.CharField(max_length=100, unique=True)
     raw_intranet_groups = models.CharField(max_length=2000)
+    affiliation = models.CharField(max_length=2000, default='')
+    unscoped_affiliation = models.CharField(max_length=2000, default='')
+    # Ideally we'd mandate mail to be unique in the database but as we already have rows in the table, we won't know
+    # what value to put in the existing rows during the migration.
+    mail = models.CharField(max_length=100, default='')
```
Comment (greenfrogs) [inline backend/uclapi/dashboard/models.py:32]: Can we have a comment explaining what mail is and the difference to email?

File: backend/uclapi/oauth/app_helpers.py [hunk 2]
```diff
@@ -21,12 +26,99 @@ def generate_random_verification_code():
 def get_student_by_upi(upi):
     """Returns a StudentA or StudentB object by UPI"""
     students = StudentsA \
```
Comment (greenfrogs) [inline backend/uclapi/oauth/app_helpers.py:28]: Can we make this one line now?

**Review by greenfrogs — COMMENTED** [review]
The changes look good

**Review by greenfrogs — APPROVED** [review]
Looks good to me

**Review by zipy124 — APPROVED** [review]

**Pull request merged**